src/threads_mngr.c

```diff
@@ -0,0 +1,187 @@
+
+/*============================ API functions implementations ========================== */
+
+void ThreadsManager_init(void) {
+    /* Register signal handler */
+    struct sigaction act;
+    sigemptyset(&act.sa_mask);
+    /* Not setting SA_RESTART flag means that If a signal handler is invoked while a
+    system call or library function call is blocked, use the default behavior
+    i.e., the call fails with the error EINTR */
+    act.sa_flags = 0;
+    act.sa_handler = invoke_callback;
+    sigaction(SIGUSR2, &act, NULL);
+}
+
+void **ThreadsManager_runOnThreads(pid_t *tids, size_t tids_len, run_on_thread_cb callback) {
+    /* Check if it is safe to start running. If not - return */
+    if(test_and_start() == IN_PROGRESS) {
+        return NULL;
+    }
+
+    /* Update g_callback */
+    g_callback = callback;
+
+    /* Set g_tids_len  */
+    g_tids_len = tids_len;
+
+    /* Allocate the output buffer */
+    g_output_array = zmalloc(sizeof(void*) * tids_len);
+
+    /* Initialize a semaphore that we will be waiting on for the threads
+    use pshared = 0 to indicate the semaphore is shared between the process's threads (and not between processes),
+    and value = 0 as the initial semaphore value.*/
+    sem_init(&wait_for_threads_sem, 0, 0);
+
+    /* Send signal to all the threads in tids */
+    pid_t pid = getpid();
+    for (size_t i = 0; i < tids_len ; ++i) {
+        syscall(SYS_tgkill, pid, tids[i], THREADS_SIGNAL);
+    }
+
+    /* Wait for all the threads to write to the output array, or until timeout is reached */
+    wait_threads();
+
+    void **ret = g_output_array;
+
+    /* Cleanups to allow next execution */
+    ThreadsManager_cleanups();
+
+    return ret;
+}
+
+/*============================ Internal functions implementations ========================== */
+
+
+static int test_and_start(void) {
+    /* atomicFlagGetSet sets the variable to 1 and returns the previous value */
+    int prev_state;
+    atomicFlagGetSet(g_in_progress, prev_state);
+
+    /* If prev_state is 1, g_in_progress was on. */
+    return prev_state;
+}
+
+static void invoke_callback(int sig) {
+    UNUSED(sig);
+
+    size_t thread_id;
+    atomicGetIncr(g_thread_ids, thread_id, 1);
+    g_output_array[thread_id] = g_callback();
+    size_t curr_done_count;
+    atomicIncrGet(g_num_threads_done, curr_done_count, 1);
+
+    /* last thread shuts down the light  */
+    if (curr_done_count == g_tids_len) {
+        sem_post(&wait_for_threads_sem);
+    }
+}
+
+static void wait_threads(void) {
+    struct timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+
+    /* calculate relative time until timeout */
+    ts.tv_sec += RUN_ON_THREADS_TIMEOUT;
+
+    /* lock the semaphore until the semaphore value rises above zero or a signal
+    handler interrupts the call. In the later case continue to wait. */
+    while ((sem_timedwait(&wait_for_threads_sem, &ts)) == -1 && errno == EINTR) {
+        continue;
+    }
```

Comment:
maybe we want to print something on timeout?, and maybe also list the threads that didn't respond?

Reply:
I don't think that there is a simple way to get such a list. we can print the requested tids list and if needed manually compare to the output tids.

Reply:
so at least print that we bailed on timeout